include/ProgramRepresentation/PVAliasSet.h

```diff
@@ -0,0 +1,71 @@
+#ifndef PV_ALIAS_SET_H
+#define PV_ALIAS_SET_H
+
+#include "ProgramRepresentation/MethodsSet.h"
+#include "ProgramRepresentation/ProgramVariable.h"
+#include "Utils.h"
+
+// holds a set of program variables and methods called on them.
+// each program variable in the set is considered an alias set.
+class PVAliasSet {
+private:
+  std::list<ProgramVariable> programVariables;
+
+  // holds the methods called on the set
+  MethodsSet methods;
+
+public:
+  PVAliasSet();
+
+  // returns true iff programVar is in the set by checking its raw name
+  bool contains(ProgramVariable programVar);
+
+  // returns true iff there exists a program variable in our set with raw name
+  // equal to rawName
+  bool containsByRawName(const std::string &rawName);
+
+  // returns true iff there exists a program variable in our set with clean name
+  // equal to cleanedName
+  bool containsByCleanedName(const std::string &cleanedName);
+
+  // adds programVar to our set of program variables. programVar is checked to
+  // see if it already exists in our set of program variables and it's not added
+  // if it already does.
+  void add(ProgramVariable programVar);
+
+  // adds array of program variables to our set of program variables.
+  // each variable is checked to see if already exists in our list of program
+  // variables.
+  void addProgramVariables(std::list<ProgramVariable> programVariables);
+
+  std::list<ProgramVariable> getProgramVariables();
+
+  MethodsSet getMethodsSet() const;
+  MethodsSet *getMethodsSetRef();
+
+  void setMethodsSet(MethodsSet methods);
+
+  // generates the set of program variables represented by their cleaned name as
+  // a string
+  std::string getVarsStringUsingCleanedNames() const;
+
+  // generates the set of program variables represented by their raw name as a
+  // string
+  std::string getVarsStringUsingRawNames() const;
+
+  // generates the set of methods called as a string.
+  std::string getMethodsString() const;
+
+  // iterates through our list of program variables to see if any of them
+  // contain an index. the first program variable we find that has an index is
+  // returned. if no such program variables exist, -1 is returned.
+  int getIndex();
```

Comment:
I put this API in because there was a bit of a tricky design decision. 

Sometimes, we create a new alias set with a program variable that refers to the field of some struct (i.e., some for some `struct my_struct { char* x, char* y }` and var `my_struct M`, we may make a new alias set `{ M.x }`. When creating that program variable, we call the `ProgramVariable(Value *value, int index)` constructor, letting us specify that this variable has a field. 

Problems arise by the fact that when we need to add more program variables to the alias set, they likely won't have a field specified. E.g., we can let some `my_struct* p = &M.x`, and it would not make sense to copy the index of `M` to `p`. There are also IR intermediate variables that complicate this as well. 

So say we have some alias set { `M.x`, `p`, `%27`, `%28` }. Of this set, only `M.x` has an index. We need to be able to access this index somehow, anyhow. (Note: a very important use of getIndex is when we look for annotations that specify an index). Since these are must-aliases, it's safe to grab any index in this set as they should all refer to `M.x` , and so, we can grab the first one which has an index.

Note that this would also work if we did something like `M a; M b; a.x = b.y`. It would be safe to grab the index of either of these variables since their annotation methods specified will be the same, or the lattice of `a.x` would be lower than `a.y`. If they aren’t, the annotation verifier will (once implemented) throw an error anyhow.

Reply:
Thanks for the explanation - this clarifies it for me. Can you add this explanation to the code as a comment? (I know it'll be long!)